man/ap.man
.TH AP %manext8% 2013-12-05 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
ap \- nmh parser of RFC 822-style addresses
.SH SYNOPSIS
.HP 5
.na
.B %nmhlibexecdir%/ap
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-form
.IR formatfile ]
.RB [ \-format
.IR string ]
.RB [ \-width
.IR columns ]
.I addrs
\&...
.ad
.SH DESCRIPTION
.B ap
is a program that parses addresses according to the ARPA
Internet standard.  It also understands many non-standard formats.
It is useful for seeing how
.B nmh
will interpret an address.
.PP
The
.B ap
program treats each argument as one or more addresses, and
prints those addresses out in the official RFC 822 format.  Hence, it is
usually best to enclose each argument in quotes for the shell.
.PP
To override the output format used by
.BR ap ,
the
.B \-format
.I string
or
.B \-format
.I file
switches are used.  This permits individual fields of
the address to be extracted with ease.  The string is simply a format
string, and the file is simply a format file.  See
.IR mh\-format (5)
for the details.
.PP
In addition to the standard escapes,
.B ap
also recognizes the following additional escape:
.PP
.RS 5
.nf
.ta \w'Escape  'u +\w'Returns  'u
.I "Escape	Returns	Description"
error	string	A diagnostic if the parse failed
.RE
.fi
.PP
Here is the default format string used by
.BR ap :
.PP
.RS 5
%<{error}%{error}: %{text}%|%(putstr(proper{text}))%>
.RE
.PP
which says that if an error was detected, print the error, a `:', and
the address in error.  Otherwise, output the RFC 822\-proper format of
the address.
.SH FILES
.PD 0
.TP 20
$HOME/.mh_profile
The user's profile.
.TP
%nmhetcdir%/mts.conf
The mts configuration file.
.PD
.SH "PROFILE COMPONENTS"
None
.SH "SEE ALSO"
.IR fmttest (1),
.IR dp (8)
.PP
.I
Standard for the Format of ARPA Internet Text Messages
(RFC 822)
.SH DEFAULTS
.PD 0
.TP 20
-format
As described above.
.TP
\-width
The width of the terminal.
.PD
.SH CONTEXT
None


NEWS
#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/ap.c
/* ap.c -- parse addresses 822-style
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/addrsbr.h>
#include <h/fmt_scan.h>
#include <h/mts.h>
#include <h/utils.h>

#define	NADDRS	100

#define	WIDTH	78

#define	FORMAT	"%<{error}%{error}: %{text}%|%(putstr(proper{text}))%>"

#define AP_SWITCHES \
    X("form formatfile", 0, FORMSW) \
    X("format string", 5, FMTSW) \
    X("width columns", 0, WIDTHSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(AP);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(AP, switches);
#undef X

static struct format *fmt;

static int dat[5];

/*
 * static prototypes
 */
static int process (char *, int);


int
main (int argc, char **argv)
{
    int addrp = 0;
    int width = -1, status = 0;
    char *cp, *form = NULL, *format = NULL, *nfs;
    char buf[BUFSIZ], **argp;
    char *addrs[NADDRS + 1]; /* Includes terminating NULL. */

    if (nmh_init(argv[0], 2)) { return 1; }

    mts_init ();

    argp = getarguments (invo_name, argc, argv, 1);
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);

		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] addrs ...",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version (invo_name);
		    done (0);

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    format = NULL;
		    continue;
		case FMTSW: 
		    if (!(format = *argp++) || *format == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    form = NULL;
		    continue;

		case WIDTHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    width = atoi (cp);
		    continue;
	    }
	}
	if (addrp == NADDRS)
	    adios (NULL, "more than %d addresses", NADDRS);
        addrs[addrp++] = cp;
    }
    addrs[addrp] = NULL;

    if (addrp == 0)
	adios (NULL, "usage: %s [switches] addrs ...", invo_name);

    /* get new format string */
    nfs = new_fs (form, format, FORMAT);

    if (width == -1) {
	if ((width = sc_width ()) < WIDTH / 2) {
	    /* Default:  width of the terminal, but at least WIDTH/2. */
	    width = WIDTH / 2;
	}
	width -= 2;
    } else if (width == 0) {
	/* Unlimited width.  */
	width = INT_MAX;
    }
    fmt_compile (nfs, &fmt, 1);

    dat[0] = 0;
    dat[1] = 0;
    dat[2] = 0;
    dat[3] = width;
    dat[4] = 0;

    for (addrp = 0; addrs[addrp]; addrp++)
	status += process (addrs[addrp], width);

    fmt_free (fmt, 1);
    done(min(status, MAX_EXIT));
    return 1;
}

struct pqpair {
    char *pq_text;
    char *pq_error;
    struct pqpair *pq_next;
};


static int
process (char *arg, int length)
{
    int	status = 0;
    char *cp;
    char error[BUFSIZ];
    struct comp *cptr;
    struct pqpair *p, *q;
    struct pqpair pq;
    struct mailname *mp;

    (q = &pq)->pq_next = NULL;
    while ((cp = getname (arg))) {
	NEW0(p);
	if ((mp = getm (cp, NULL, 0, error, sizeof(error))) == NULL) {
	    p->pq_text = mh_xstrdup(cp);
	    p->pq_error = mh_xstrdup(error);
	    status++;
	}
	else {
	    p->pq_text = getcpy (mp->m_text);
	    mnfree (mp);
	}
	q = (q->pq_next = p);
    }

    for (p = pq.pq_next; p; p = q) {
	charstring_t scanl =
	    charstring_create (length < NMH_BUFSIZ ? length : NMH_BUFSIZ);

	cptr = fmt_findcomp ("text");
	if (cptr) {
            mh_xfree(cptr->c_text);
	    cptr->c_text = p->pq_text;
	    p->pq_text = NULL;
	}
	cptr = fmt_findcomp ("error");
	if (cptr) {
            mh_xfree(cptr->c_text);
	    cptr->c_text = p->pq_error;
	    p->pq_error = NULL;
	}

	fmt_scan (fmt, scanl, length, dat, NULL);
	fputs (charstring_buffer (scanl), stdout);
	charstring_free (scanl);

        mh_xfree(p->pq_text);
        mh_xfree(p->pq_error);
	q = p->pq_next;
	free(p);
    }

    return status;
}
